### ISSUE TITLE
spring boot startup performance issue when large properties file and spring-cloud-config-server dependency

### ISSUE THREAD

**Issue by orenw1** [spring-projects/spring-boot#10696]
Scenario:
	Spring Boot web application (1.5.2.RELEASE) using @PropertySource point to properties
        file that contain over 5000 properties.
	Without spring-cloud-config-server dependency and with the large properties file startup time: ~8 seconds
	With adding dependency of spring-cloud-config-server (1.2.3.RELEASE), startup time takes ~ 30 seconds
	With adding dependency of spring-cloud-config-server with only 1 property in this file it takes ~10 seconds
	Looks like large properties number + spring-cloud-config-server increase a lot 
       the startup time.
	Profiling show: 
       org.springframework.boot.bind.PropertySourcesPropertyValues.processEnumerablePropertySource()
        Made the same tests with  Spring Boot 2.0.0.M5, spring-cloud-config-server 2.0.0.M2
	There is significant improvement, with the large file and dependency: ~12 seconds

1. are you aware of it?
2. is there a plan to enhance also 1.5.x version?

thanks,

   Oren.

**Comment by wilkinsona** [comment]
Given the move from 8 seconds to 30 seconds when you add Spring Cloud Config Server, this sounds like a Spring Cloud issue to me, although it's hard to tell as for the improved time (12 seconds), you've changed both the version of Boot and the version of Spring Cloud Config Server.

To answer your questions:

> are you aware of it?

We're not aware of any dramatic improvements in binding performance between Boot 1.5 and 2.0.

> is there a plan to enhance also 1.5.x version?

It's not clear that the enhancement was in Boot.

I'd recommend discussing your finding with the Spring Cloud team on [their Gitter channel](https://gitter.im/spring-cloud/spring-cloud) and we can then take things from there.

**Comment by orenw1** [comment]
thanks, i am asking it also in Gitter, just saw this pull request and thought that might be related
[Alternative implementation of PropertyNamePatternsMatcher](https://github.com/spring-projects/spring-boot/pull/3402/files?diff=split)
also could not test with the older spring-cloud-config-server as it failed in runtime.

**Comment by wilkinsona** [comment]
@orenw1 That change is unrelated. It was made in 1.3 and you are comparing 1.5 and 2.0.

**Comment by orenw1** [comment]
hi,
  i got the below response in Spring-cloud Gitter channel:

   "Dave Syer @dsyer Oct 28 19:40
   I did some analysis. I don't think it's anything to do with Spring Cloud.
   You just have more beans in a cloud app
   So the effect is magnified
   You can open an issue in Spring Boot if you like. It might not be possible to fix it in 1.5.x."

  there is also an example i created to demonstrate the scenario:
  [large_properties_file_example](https://github.com/orenw1/large_properties_file_example)
  
  Can you evaluate if it is something that can be handled in 1.5.x version?

thanks

**Comment by wilkinsona** [comment]
Thanks for the example. Unfortunately, I don't think there's anything that we can do about this in 1.5.x. It's an unfortunate side-effect of 1.x's brute force approach to relaxed binding that requires a fairly large number of combinations to be compared.

I've profiled the app running with short key names and with long key names in YourKit. The latter shows `org.springframework.boot.bind.DefaultPropertyNamePatternsMatcher.matches(String)` has a hot spot whereas the former does not. The longer property prefixes are giving it more work to do as there are more characters to compare. The long names also prevent short-circuiting when the names that are being matched are longer than the property names.

The `matches(String)` method has already been optimised for frequent execution and I can't see anything obvious to make it any faster. The few things that I have tried have had no effect at best. I think we'll have to close this one knowing that things are much improved already in 2.0.
